tfjs-core/src/ops/linalg_ops.ts

```diff
@@ -22,13 +22,109 @@
 import {ENGINE} from '../engine';
 import {dispose} from '../globals';
 import {Tensor, Tensor1D, Tensor2D} from '../tensor';
+import {convertToTensor} from '../tensor_util_env';
+import {TensorLike} from '../types';
 import {assert} from '../util';
 import {eye, squeeze, stack, unstack} from './array_ops';
+import {sub} from './binary_ops';
 import {split} from './concat_split';
+import {logicalAnd, where} from './logical_ops';
 import {norm} from './norm';
 import {op} from './operation';
 import {sum} from './reduction_ops';
-import {tensor2d} from './tensor_ops';
+import {range, scalar, tensor2d, zeros} from './tensor_ops';
+
+/**
+ * Copy a tensor setting everything outside a central band in each innermost
+ * matrix to zero.
+ *
+ * The band part is computed as follows: Assume input has `k` dimensions
+ * `[I, J, K, ..., M, N]`, then the output is a tensor with the same shape where
+ * `band[i, j, k, ..., m, n] = in_band(m, n) * input[i, j, k, ..., m, n]`.
+ * The indicator function
+ * `in_band(m, n) = (num_lower < 0 || (m-n) <= num_lower))`
+ * `&& (num_upper < 0 || (n-m) <= num_upper)`
+ *
+ * ```js
+ * const x = tf.tensor2d([[ 0,  1,  2, 3],
+ *                        [-1,  0,  1, 2],
+ *                        [-2, -1,  0, 1],
+ *                        [-3, -2, -1, 0]]);
+ * let y = tf.linalg.bandPart(x, 1, -1);
+ * y.print(); // [[ 0,  1,  2, 3],
+ *            //  [-1,  0,  1, 2],
+ *            //  [ 0, -1,  0, 1],
+ *            //  [ 0, 0 , -1, 0]]
+ * let z = tf.linalg.bandPart(x, 2, 1);
+ * z.print(); // [[ 0,  1,  0, 0],
+ *            //  [-1,  0,  1, 0],
+ *            //  [-2, -1,  0, 1],
+ *            //  [ 0, -2, -1, 0]]
+ * ```
+ *
+ * @param x Rank `k` tensor
+ * @param numLower Number of subdiagonals to keep.
+ *   If negative, keep entire lower triangle.
+ * @param numUpper Number of subdiagonals to keep.
+ *   If negative, keep entire upper triangle.
+ * @returns Rank `k` tensor of the same shape as input.
+ *   The extracted banded tensor.
+ */
+/**
+ * @doc {heading:'Operations',
+ *       subheading:'Linear Algebra',
+ *       namespace:'linalg'}
+ */
+function bandPart_<T extends Tensor>(
+  a: T|TensorLike, numLower: number, numUpper: number
+): T
+{
+  if( numLower%1 !== 0 ){
+    throw new Error(`bandPart(): numLower=${numLower} not an integer.`);
+  }
+  if( numUpper%1 !== 0 ){
+    throw new Error(`bandPart(): numUpper=${numUpper} not an integer.`);
+  }
+
+  return ENGINE.tidy( () => {
+    const $a = convertToTensor(a,'a','bandPart');
+    a = undefined;
```

Comment:
The idea was to prevent myself from using `a` instead of `$a` in the remaining lines of code. It was meant for assertion purposes only and does not belong in the final commit. I'm going to remove it.

Reply:
If I am not mistaken, the `ENGINE.tidy` is redundant, right? `op({_bandPart})` already adds a scope which disposes the local tensors.

Reply:
Yes, the `ENGINE.tidy` would have been redundant. Thanks!